domain/consensus/processes/consensusstatemanager/verify_and_build_utxo.go

```diff
@@ -0,0 +1,109 @@
+package consensusstatemanager
+
+import (
+	"sort"
+
+	"github.com/kaspanet/kaspad/domain/consensus/utils/hashserialization"
+
+	"github.com/kaspanet/kaspad/domain/consensus/utils/merkle"
+
+	"github.com/kaspanet/kaspad/domain/consensus/utils/hashes"
+
+	"github.com/kaspanet/kaspad/domain/consensus/model"
+	"github.com/kaspanet/kaspad/domain/consensus/model/externalapi"
+	"github.com/kaspanet/kaspad/domain/consensus/ruleerrors"
+	"github.com/pkg/errors"
+)
+
+func (csm *consensusStateManager) verifyAndBuildUTXO(block *externalapi.DomainBlock, blockHash *externalapi.DomainHash,
+	pastUTXODiff *model.UTXODiff, acceptanceData []*model.BlockAcceptanceData, multiset model.Multiset) error {
+
+	err := csm.validateUTXOCommitment(block, blockHash, multiset)
+	if err != nil {
+		return err
+	}
+
+	err = csm.validateAcceptedIDMerkleRoot(block, blockHash, acceptanceData)
+	if err != nil {
+		return err
+	}
+
+	err = csm.validateBlockTransactionsAgainstPastUTXO(block, blockHash, pastUTXODiff, err)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (csm *consensusStateManager) validateBlockTransactionsAgainstPastUTXO(block *externalapi.DomainBlock,
+	blockHash *externalapi.DomainHash, pastUTXODiff *model.UTXODiff, err error) error {
+
+	ghostdagData, err := csm.ghostdagDataStore.Get(csm.databaseContext, blockHash)
+	if err != nil {
+		return err
+	}
+	selectedParentMedianTime, err := csm.pastMedianTimeManager.PastMedianTime(ghostdagData.SelectedParent)
+	if err != nil {
+		return err
+	}
+
+	for _, transaction := range block.Transactions {
+		err = csm.populateTransactionWithUTXOEntriesFromVirtualOrDiff(transaction, pastUTXODiff)
```

Comment:
You are already calling this in maybeAcceptTransaction. Make sure there's no duplication here.

Reply:
maybeAcceptTransaction is for transactions in newBlock.blues. validateBlockTransactionsAgainstPastUTXO (a.k.a. this guy) is for transactions in the new block.
Therefore, no duplication